arraycontext/pytest.py

```diff
@@ -28,48 +29,124 @@
 THE SOFTWARE.
 """
 
+from typing import Optional, Sequence
 
-# {{{ pytest integration
+import pyopencl as cl
+from pyopencl.tools import _ContextFactory
 
-def pytest_generate_tests_for_pyopencl_array_context(metafunc):
-    """Parametrize tests for pytest to use a
-    :class:`~arraycontext.PyOpenCLArrayContext`.
 
-    Performs device enumeration analogously to
-    :func:`pyopencl.tools.pytest_generate_tests_for_pyopencl`.
+# {{{ array context factories
 
-    Using the line:
+class _PyOpenCLArrayContextFactory(_ContextFactory):
+    force_device_scalars = True
+
+    def __call__(self):
+        ctx = super().__call__()
+        from arraycontext.impl.pyopencl import PyOpenCLArrayContext
+        return PyOpenCLArrayContext(
+                cl.CommandQueue(ctx),
+                force_device_scalars=self.force_device_scalars)
+
+    def __str__(self):
+        return ("<array context factory for <pyopencl.Device '%s' on '%s'>" %
+                (self.device.name.strip(),
+                 self.device.platform.name.strip()))
+
+
+class _DeprecatedPyOpenCLArrayContextFactory(_PyOpenCLArrayContextFactory):
+    force_device_scalars = False
+
+
+_ARRAY_CONTEXT_FACTORY_DICT = {
+        "pyopencl": _PyOpenCLArrayContextFactory,
+        }
+_ALL_ARRAY_CONTEXT_FACTORY_DICT = {
+        "pyopencl-deprecated": _DeprecatedPyOpenCLArrayContextFactory,
+        }
+
+_ALL_ARRAY_CONTEXT_FACTORY_DICT.update(_ARRAY_CONTEXT_FACTORY_DICT)
+
+# }}}
+
+
+# {{{ pytest integration
+
+def pytest_generate_tests_for_array_context(
```

Comment:
```suggestion
def pytest_generate_tests_for_array_contexts(
```
?

Reply:
I kept it singular because the pyopencl iss singular even though it generates a bunch of variants. This should read better in light of https://github.com/alexfikl/arraycontext/pull/1#discussion_r657501961, so will do!

Reply:
Renamed in 80d5e4b.